hkmc2/shared/src/main/scala/hkmc2/codegen/HandlerLowering.scala

```diff
@@ -0,0 +1,561 @@
+  
+  def partitionBlock(blk: Block, labelIds: Map[Symbol, (StateId, StateId)] = Map.empty): Ls[BlockState] = 
+    // for readability :)
+    case class PartRet(head: Block, states: Ls[BlockState])
+
+    // * returns (truncated input block, child block states)
+    // * blk: The block to transform
+    // * labelIds: maps label IDs to the state at the start of the label and the state after the label
+    // * jumpTo: what state End should jump to, if at all 
+    // * freshState: uid generator
+    // TODO: don't split within Match, Begin and Labels when not needed, ideally keep it intact. Need careful analysis for this.
+    def go(blk: Block)(implicit labelIds: Map[Symbol, (StateId, StateId)], afterEnd: Option[StateId]): PartRet = blk match
+      case ResumptionPoint(result, uid, rest) =>
+        val PartRet(head, states) = go(rest)
+        PartRet(StateTransition(uid), BlockState(uid, head, S(result)) :: states)
+
+      case Match(scrut, arms, dflt, rest) => 
+        val restParts = go(rest)
+        val restId: StateId = restParts.head match
+          case StateTransition(uid) => uid
+          case _ => freshId()
+        
+        val armsParts = arms.map((cse, blkk) => (cse, go(blkk)(afterEnd = S(restId))))
+        val dfltParts = dflt.map(blkk => go(blkk)(afterEnd = S(restId)))
+        
+
+        val states_ = restParts.states ::: armsParts.flatMap(_._2.states)
+        val states = dfltParts match
+          case N => states_
+          case S(value) => value.states ::: states_
+
+        val newArms = armsParts.map((cse, partRet) => (cse, partRet.head))
+        
+        restParts.head match
+          case StateTransition(_) =>
+            PartRet(
+              Match(scrut, newArms, dfltParts.map(_.head), StateTransition(restId)),
+              states
+            )
+          case _ =>
+            PartRet(
+              Match(scrut, newArms, dfltParts.map(_.head), StateTransition(restId)),
+              BlockState(restId, restParts.head, N) :: states
+            )
+      case l @ Label(label, body, rest) =>
+        val startId = freshId() // start of body
+
+        val PartRet(restNew, restParts) = go(rest)
+
+        val endId: StateId = restNew match // start of rest
+          case StateTransition(uid) => uid 
+          case _ => freshId()
+
+        val PartRet(bodyNew, parts) = go(body)(using labelIds + (label -> (startId, endId)), S(endId))
+        
+        restNew match
+          case StateTransition(_) =>
+            PartRet(
+              StateTransition(startId), 
+              BlockState(startId, bodyNew, N) :: parts ::: restParts
+            )
+          case _ =>
+            PartRet(
+              StateTransition(startId), 
+              BlockState(startId, bodyNew, N) :: BlockState(endId, restNew, N) :: parts ::: restParts
+            )
+
+      case Break(label) =>
+        val (start, end) = labelIds.get(label) match
+          case N => raise(ErrorReport(
```

Comment:
This looks likes it should never happen unless there's a compiler bug, isn't it? In this case you should just crash the compiler.

```suggestion
          case N => raise(InternalError(
```

`raise` is for actionable _user_ errors with nice diagnostic messages.

Reply:
> `raise` is for actionable user errors with nice diagnostic messages.
I suppose that's unless an `InternalError` is raised, as shown above?

Reply:
Ah yeah I that's what I meant.